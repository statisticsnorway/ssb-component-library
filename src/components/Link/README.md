Link
========

> Different type of links
### Usage


#### HTML

```html
<a class="ssb-link" href=" " target="" rel="">
    <span class="link-text">A regular link</span>
</a>

<a class="ssb-link with-icon" href=" " target="" rel="">
    <div class="icon-wrapper">
        <i>{feather.arrowRight 16px}</i>
    </div>
    <span class="link-text">A regular link with icon</span>
</a>

<a class="ssb-link profiled mr-4" href=" " target="" rel="">
    <span class="link-text">A profiled link</span>
</a>

<a class="ssb-link profiled with-icon" href=" " target="" rel="">
    <div class="icon-wrapper">
        <i>{feather.arrowRight 20px}</i>
    </div>
    <span class="link-text">A profiled link with icon</span>
</a>

<a class="ssb-link header" href=" " target="" rel="">
    <span class="link-text">A header link</span>
</a>

<a class="ssb-link negative" href=" " target="" rel="">
    <span class="link-text">A regular link - Negative</span>
</a>
```

#### React

```jsx harmony
<Link href="">A regular link</Link>

<Link href="" icon={<ArrowRight size={16} />}>A regular link with icon</Link>

<Link href="" linkType="profiled">A profiled link</Link>

<Link href=" " linkType="profiled" icon={<ArrowRight size={20} />}>A profiled link with icon</Link>

<Link href=" " linkType="header">A header link</Link>

<Link href="" negative>A regular link - Negative</Link>


```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| children   | Required node | Something that renders to wrap the link around |
| className   | string | Optional container class|
| href | Required string| Destination for navigation |
| icon | Optional node | Renders an icon  |
| isExternal | Optional boolean | Will make the link open in new tab. Defaults to false |
| linkType | Optional oneOf |Changes the style. Can be 'regular', 'profiled', or 'header'. Defaults to 'regular' |
| negative | Optional boolean | Will change component style. Defaults to false |
