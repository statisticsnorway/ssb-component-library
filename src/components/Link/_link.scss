$default-font-size: 16px;
$profiled-font-size: 18px;
$header-font-size: 20px;
$icon-padding: 8px;

.ssb-link {
  @include open-sans;
  align-items: center;
  background-image: linear-gradient(120deg, $ssb-green-4 0%, $ssb-green-4 100%);
  background-position: 0 100%;
  background-repeat: no-repeat;
  background-size: 100% 0;
  border-bottom: 1px solid $ssb-green-4;
  color: $ssb-green-4;
  cursor: pointer;
  display: inline;
  font-size: $default-font-size;
  line-height: 1.7;
  margin-top: -3px;
  margin-bottom: -2px;
  padding: 0 1px 2px;
  position: relative;
  text-decoration: none;
  transition:
    background-size 0.2s ease-in,
    color 0.1s;

  .icon-wrapper {
    display: flex;
    margin-right: 4px;
    margin-top: 5px;
  }

  &.header {
    @include roboto;
    border-bottom-width: 2px;
    font-size: $header-font-size;
    font-weight: bold;

    .link-text {
      color: $ssb-dark-6;
    }
  }

  &:hover,
  &:focus {
    background-size: 100% 100%;
    color: $ssb-white;
    outline: 0;
    text-decoration: none;

    .link-text {
      color: $ssb-white;
    }
  }

  &.negative {
    background-image: linear-gradient(120deg, $ssb-green-2 0%, $ssb-green-2 100%);
    border-bottom: 1px solid $ssb-green-2;
    color: $ssb-green-2;

    .link-text {
      color: $ssb-green-2;
    }

    &.header {
      border-bottom-width: 2px;

      .link-text {
        color: $ssb-white;
      }
    }

    &.with-icon {
      padding: 2px 2px 0;
      border-bottom: none;
      display: inline-flex;
      align-items: flex-start;

      i {
        font-size: 20px;
      }

      .link-text {
        color: $ssb-white;
      }

      &:hover,
      &:focus {
        color: $ssb-dark-5;

        .link-text {
          color: $ssb-dark-5;
        }
      }
    }

    &:hover,
    &:focus {
      color: $ssb-dark-5;
      outline: 0;

      .link-text {
        color: $ssb-dark-5;
      }
    }
  }

  &.profiled {
    border-bottom-width: 2px;
    font-size: $profiled-font-size;
    font-weight: bold;
  }

  &.with-icon {
    padding: 2px 2px 0;
    border-bottom: none;
    display: inline-flex;
    align-items: flex-start;

    i {
      font-size: 16px;
    }

    .link-text {
      color: $ssb-dark-6;
    }

    &:hover,
    &:focus {
      color: $ssb-white;

      .link-text {
        color: $ssb-white;
      }
    }
  }

  .link-text {
    color: $ssb-green-4;
    transition: color 0.1s;

    &:hover,
    &:focus {
      color: $ssb-white;
      outline: 0;
    }
  }

  &.stand-alone {
    min-height: 44px;
    min-width: 44px;
    display: inline-flex;
    align-items: center !important;
    border-bottom: none;
    text-decoration: underline;
    text-underline-offset: 0.3em;

    .icon-wrapper {
      margin-top: 0;
    }

    &.with-icon {
      text-decoration: none;
    }

    &:hover {
      text-decoration: none;
    }
  }
}
