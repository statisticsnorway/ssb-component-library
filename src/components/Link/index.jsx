import React from 'react';
import PropTypes from 'prop-types';

const Link = ({
	ariaLabel, children, className, href, icon, isExternal, linkType, negative, tabIndex, title,
}) => {
	const classNames = `ssb-link${linkType ? ` ${linkType}` : ''}${negative ? ' negative' : ''}${icon ? ' with-icon' : ''}${className ? ` ${className}` : ''}`;

	return (
		<a
			className={classNames}
			href={href}
			target={isExternal ? '_blank' : undefined}
			rel={isExternal ? 'noopener noreferrer' : undefined}
			tabIndex={tabIndex}
			aria-label={ariaLabel}
			title={title}
		>{icon && <div className="icon-wrapper">{icon}</div>}
			{children && <span className="link-text">{children}</span>}
		</a>
	);
};

Link.defaultProps = {
	className: '',
	isExternal: false,
	negative: false,
};

Link.propTypes = {
	ariaLabel: PropTypes.string,
	children: PropTypes.node,
	className: PropTypes.string,
	href: PropTypes.string.isRequired,
	icon: PropTypes.node,
	isExternal: PropTypes.bool,
	linkType: PropTypes.oneOf([
		'regular',
		'profiled',
		'header',
	]),
	negative: PropTypes.bool,
	tabIndex: PropTypes.number,
	title: PropTypes.string,
};

export default Link;
