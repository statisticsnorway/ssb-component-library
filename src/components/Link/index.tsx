import React, { createElement, forwardRef, type ReactNode } from 'react'

interface LinkProps {
  ariaLabel?: string
  children?: ReactNode
  className?: string
  href: string
  icon?: ReactNode
  isExternal?: boolean
  linkType?: 'regular' | 'profiled' | 'header'
  negative?: boolean
  tabIndex?: number
  title?: string
  onClick?: () => void
  standAlone?: boolean
  headingSize?: 1 | 2 | 3 | 4 | 5 | 6
}

const Link = forwardRef<HTMLAnchorElement, LinkProps>(
  (
    {
      ariaLabel,
      children,
      className,
      href,
      icon,
      isExternal = false,
      linkType,
      negative = false,
      tabIndex,
      title,
      onClick = () => {},
      standAlone,
      headingSize,
    },
    ref
  ) => {
    const classNames = `ssb-link${linkType ? ` ${linkType}` : ''}${standAlone ? ' stand-alone' : ''}
    ${negative ? ' negative' : ''}${icon ? ' with-icon' : ''}${className ? ` ${className}` : ''}`

    const renderLinkText = () => {
      if (children) {
        if (linkType === 'header' && headingSize) {
          return createElement(`h${headingSize}`, { className: 'link-text' }, children)
        }
        return <span className='link-text'>{children}</span>
      }
      return null
    }

    return (
      <a
        className={classNames.replace(/\s\s+/g, ' ').trim()}
        href={href}
        target={isExternal ? '_blank' : undefined}
        rel={isExternal ? 'noopener noreferrer' : undefined}
        tabIndex={tabIndex}
        aria-label={ariaLabel}
        title={title}
        onClick={onClick}
        ref={ref}
      >
        {icon && <div className='icon-wrapper'>{icon}</div>}
        {renderLinkText()}
      </a>
    )
  }
)

Link.displayName = 'Link'

export default Link
