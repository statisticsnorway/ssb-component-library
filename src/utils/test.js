import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
