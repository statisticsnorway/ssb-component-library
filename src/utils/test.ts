import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
