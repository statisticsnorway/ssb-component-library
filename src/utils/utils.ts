// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
