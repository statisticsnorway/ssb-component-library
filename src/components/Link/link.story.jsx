import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import { ArrowRight } from 'react-feather';
import Link from './index';

const negativeStyle = {
	backgroundColor: '#274247',
	padding: '100px',
};

storiesOf('Link', module).addDecorator(centered)
	.add('Default', () => (
		<div>
			<p>Check it out a <Link href=" "> regular link</Link> within a sentence!</p>
			<Link href=" " icon={<ArrowRight size={16} />}>A regular link with icon</Link>
		</div>
	))
	.add('Profiled', () => (
		<div>
			<p><Link href=" " linkType="profiled">A profiled link</Link></p>
			<Link href=" " linkType="profiled" icon={<ArrowRight size={20} />}>A profiled link with icon</Link>
		</div>
	))
	.add('Header', () => (
		<div>
			<p><Link href=" " linkType="header">A header link</Link></p>
		</div>
	))
	.add('Negative', () => (
		<div style={negativeStyle}>
			<p><Link href=" " negative>A regular link - Negative</Link></p>
			<Link href=" " icon={<ArrowRight size={16} />} negative>A regular link with icon - Negative</Link>
			<p><Link href=" " linkType="profiled" negative>A profiled link - Negative</Link></p>
			<Link href=" " linkType="profiled" icon={<ArrowRight size={20} />} negative>A profiled link with icon - Negative</Link>
			<p><Link href=" " linkType="header" negative>A header link - Negative</Link></p>
		</div>
	));
