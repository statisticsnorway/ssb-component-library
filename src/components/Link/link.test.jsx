import React from 'react'
import { render } from '../../utils/test'
import { ArrowRight } from 'react-feather'
import Link from './index'

describe('Link component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<Link href=''>A regular link</Link>)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Toggles classNames correctly', () => {
    const { asFragment } = render(
      <Link className='mt-3' isExternal href='' negative linkType='profiled'>
        Profile link
      </Link>
    )
    expect(asFragment()).toMatchSnapshot()
  })
  test('render icon', () => {
    const { asFragment } = render(
      <Link href=' ' linkType='profiled' icon={<ArrowRight />}>
        A profiled link with icon
      </Link>
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
