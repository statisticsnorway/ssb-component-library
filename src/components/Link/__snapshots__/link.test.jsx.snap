// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link component Matches the snapshot 1`] = `
<a
  className="ssb-link"
  href=""
>
  <span
    className="link-text"
  >
    A regular link
  </span>
</a>
`;
